control-plane/agents/src/bin/ha/cluster/nodes.rs

```diff
@@ -0,0 +1,63 @@
+use crate::volume::VolumeMover;
+use std::{collections::HashMap, net::SocketAddr, sync::Arc};
+use tokio::sync::Mutex;
+use tracing::info;
+
+/// Store node information and reported failed path
+#[derive(Debug, Clone)]
+pub struct NodeList {
+    list: Arc<Mutex<HashMap<String, SocketAddr>>>,
+    failed_path: Arc<Mutex<HashMap<String, SocketAddr>>>,
+}
+
+impl NodeList {
+    pub fn new() -> Self {
+        NodeList {
+            list: Arc::new(Mutex::new(HashMap::new())),
+            failed_path: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    /// register node and its endpoint
+    /// if node is already registered then update it details
+    pub async fn register_node(&self, name: String, endpoint: SocketAddr) {
+        let mut list = self.list.lock().await;
+        list.insert(name.clone(), endpoint);
+    }
+
+    pub async fn remove_failed_path(&self, path: String) {
+        let mut failed_path = self.failed_path.lock().await;
+        failed_path.remove(&path);
+    }
+
+    /// report_failed_path send switchover request to switchover engine for the
+    /// reported node and path
+    pub async fn report_failed_path(
+        self,
+        node: String,
+        path: String,
+        mover: VolumeMover,
+    ) -> Result<(), anyhow::Error> {
+        let mut failed_path = self.failed_path.lock().await;
+
+        if failed_path.get(&path).is_some() {
+            return Err(anyhow::format_err!(
+                "Path {} is already reported for switchover",
+                path
+            ));
+        };
+
+        let list = self.list.lock().await;
```

Comment:
we're taking 2 consecutive locks here, how do we ensure we don't get into deadlocks?